napari/_vispy/vispy_base_layer.py

```diff
@@ -204,9 +290,19 @@ def on_mouse_release(self, event):
         self.layer.on_mouse_release(event)
 
     def on_draw(self, event):
-        """Called whenever the canvas is drawn.
+        """Called whenever the canvas is drawn, which happens whenever new
+        data is sent to the canvas or the camera is moved.
         """
         self.layer.scale_factor = self.scale_factor
+        if self.layer.is_pyramid:
+            self.layer.scale_factor = self.scale_factor
+            size = self.camera.rect.size
+            data_level = self.compute_data_level(size)
+
+            if data_level != self.layer.data_level:
+                self.layer.data_level = data_level
```

Comment:
comment added - it is because setting `data_level` will trigger computation that recalculates the corner pixels now that the `data_level` has changed.